Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp

```diff
@@ -0,0 +1,386 @@
+        encoder->close();
+    });
+}
+
+InternalAudioEncoderCocoa::InternalAudioEncoderCocoa(const Config& config, AudioEncoder::DescriptionCallback&& descriptionCallback, AudioEncoder::OutputCallback&& outputCallback)
+    : m_descriptionCallback(WTFMove(descriptionCallback))
+    , m_outputCallback(WTFMove(outputCallback))
+    , m_config(config)
+{
+}
+
+String InternalAudioEncoderCocoa::initialize(const String& codecName)
+{
+    auto result = AudioDecoderCocoa::isCodecSupported(codecName);
+    if (!result)
+        return result.error();
+
+    std::tie(m_codec, m_pcmFormat) = *result;
+    if (m_codec == 'vorb')
+        return "Vorbis encoding is not supported"_s;
+    if (m_codec == kAudioFormatFLAC)
+        return "FLAC encoding is not supported"_s;
+    if (m_codec == kAudioFormatMPEGLayer3)
+        return "MP3 encoding is not supported"_s;
+
+    if (m_codec == kAudioFormatOpus) {
+        if (m_config.numberOfChannels > 2)
+            return "Opus with more than two channels is not supported"_s;
+
+        if (m_config.opusConfig && m_config.opusConfig->isOggBitStream)
+            return "Opus Ogg format is unsupported"_s;
+
+        auto config = m_config.opusConfig.value_or(AudioEncoder::OpusConfig { });
+        m_frameDuration = MediaTime(config.frameDuration, 1000000);
+        m_complexity = config.complexity;
+        m_packetlossperc = config.packetlossperc;
+        m_useinbandfec = config.useinbandfec;
+    } else if ((m_codec == kAudioFormatMPEG4AAC || m_codec == kAudioFormatMPEG4AAC_HE || m_codec == kAudioFormatMPEG4AAC_LD || m_codec == kAudioFormatMPEG4AAC_HE_V2 || m_codec == kAudioFormatMPEG4AAC_ELD)) {
+        m_isADTS = m_config.isAacADTS.value_or(false);
+        if (m_isADTS)
+            return "AAC ADTS format is unsupported"_s;
+    }
+
+    if (m_pcmFormat)
+        m_outputDescription = CAAudioStreamDescription { double(m_config.sampleRate), uint32_t(m_config.numberOfChannels), *m_pcmFormat, CAAudioStreamDescription::IsInterleaved::Yes };
+    else {
+        AudioStreamBasicDescription asbd { };
+        asbd.mFormatID = m_codec;
+        asbd.mSampleRate = double(m_config.sampleRate);
+        asbd.mChannelsPerFrame = uint32_t(m_config.numberOfChannels);
+        if (m_frameDuration)
+            asbd.mFramesPerPacket = m_config.sampleRate * m_frameDuration->toDouble();
+        m_outputDescription = asbd;
+    }
+    m_codecString = codecName.isolatedCopy();
+
+    return emptyString();
+}
+
+void InternalAudioEncoderCocoa::compressedAudioOutputBufferCallback(void* object, CMBufferQueueTriggerToken)
+{
+    // We can only be called from the CoreMedia callback if we are still alive.
+    RefPtr encoder = static_cast<class InternalAudioEncoderCocoa*>(object);
+
+    InternalAudioEncoderCocoa::queueSingleton().dispatch([weakEncoder = ThreadSafeWeakPtr { *encoder }] {
+        if (auto strongEncoder = weakEncoder.get())
+            strongEncoder->processEncodedOutputs();
```

Comment:
It is not clear to me that `processEncodedOutputs()` will be executed before we resolve the encode native promise. Maybe we should keep the native promise producer somewhere and resolve/reject as part of processEncodedOutputs (and handle the case where processEncodedOutputs is not called).

Reply:
This is always called before because we return the promise itself which is resolved after the frames have returned. Same pattern as AudioDecoder

Reply:
I added a call to processEncodedOutput() at the completion of the AudioSampleBufferConverter promise to remove all doubts.